src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -618,32 +520,95 @@ def calculate(
-                inclusion_lower_bounds,
-                inclusion_upper_bounds,
-            ) = self._fetch_inclusion_bounds(battery_id, inverter_id, metrics_data)
-            timestamp = max(timestamp, _ts)
-            (
-                _ts,
-                exclusion_lower_bounds,
-                exclusion_upper_bounds,
-            ) = self._fetch_exclusion_bounds(battery_id, inverter_id, metrics_data)
-            if len(inclusion_upper_bounds) > 0:
-                inclusion_bounds_upper += min(inclusion_upper_bounds)
-            if len(inclusion_lower_bounds) > 0:
-                inclusion_bounds_lower += max(inclusion_lower_bounds)
-            if len(exclusion_upper_bounds) > 0:
-                exclusion_bounds_upper += max(exclusion_upper_bounds)
-            if len(exclusion_lower_bounds) > 0:
-                exclusion_bounds_lower += min(exclusion_lower_bounds)
+        battery_sets = {
+            self._bat_bats_map[battery_id] for battery_id in working_batteries
+        }
+
+        def get_validated_bounds(
+            comp_id: int, comp_metric_ids: list[ComponentMetricId]
+        ) -> PowerBounds | None:
+            results: list[float] = []
+            # Make timestamp accessible
+            nonlocal loop_timestamp
+            local_timestamp = loop_timestamp
+
+            if data := metrics_data.get(comp_id):
+                for comp_metric_id in comp_metric_ids:
+                    val = data.get(comp_metric_id)
+                    if val is not None:
+                        local_timestamp = max(loop_timestamp, data.timestamp)
+                        results.append(val)
+
+            if len(results) != len(comp_metric_ids):
+                return None
+
+            loop_timestamp = local_timestamp
+            return PowerBounds(
+                inclusion_lower=results[0],
+                exclusion_lower=results[1],
+                exclusion_upper=results[2],
+                inclusion_upper=results[3],
+            )
+
+        for battery_ids in battery_sets:
+            loop_timestamp = timestamp
+
+            inverter_ids = self._bat_inv_map[next(iter(battery_ids))]
+
+            battery_bounds = list(
+                x
+                for x in map(
+                    lambda battery_id: get_validated_bounds(
+                        battery_id, self._battery_metrics
+                    ),
+                    battery_ids,
+                )
+                if x is not None
+            )
+
+            if len(battery_bounds) == 0:
+                continue
+
+            aggregated_bat_bounds = _aggregate_battery_power_bounds(battery_bounds)
+
+            inverter_bounds = list(
+                x
+                for x in map(
+                    lambda inverter_id: get_validated_bounds(
+                        inverter_id, self._inverter_metrics
+                    ),
+                    inverter_ids,
+                )
+                if x is not None
+            )
```

Comment:
I wonder if the calculation of `battery_bounds` and `inverters_bounds` can be done in one helper passing battery/inverter IDs and metrics respectively,  to avoid code duplication and reduce a bit the size of the `calculate() ` function